Introduce user provider and resource

martinb3 opened the pull request:
- Fixes #269, adds an `elasticsearch_user` resource with provider that creates a user, group, and removes the automatic homedir

- Pins chef to 11.16.4 since chef 12 and many libraries like poise aren't compatible yet

- Adds an `elasticsearch_test` kitchenCI suite that tests edge cases, so the default test suite can test the expected way people use the cookbook

- Tweak rubocop rules to avoid style complaints about methods that are too complex (controversial rule) and compact class styles (also controversial at the moment)

- Updated `README.md` with an example of how to use the `elasticsearch_user` resource

martinb3 commented:
Force pushed a new commit message into my branch, to add the `fixes #XXX` to the commit message.

karmi commented on libraries/user.rb:
```diff
@@ -0,0 +1,70 @@
+  class Resource::ElasticsearchUser < Resource
+    include Poise
+
+    actions(:create, :remove)
```
What does everybody think about getting rid of the `()` parentheses for these DSL-like notations? I guess it would look & feel cleaner?

martinb3 replied:
I would recommend leaving them, since we're emulating a style that other users and chef community people will recognize. It seems like they are most often omitted in LWRPs and not omitted in HWRPs. I don't feel that strongly about it though, if everyone else feels differently. I just want to be sure what we write is approachable enough that we can get good contributions from the community.

karmi commented on libraries/user.rb:
```diff
@@ -0,0 +1,70 @@
+            not_if  { ::File.symlink?("#{computed_homedir}") }
+            only_if { ::File.directory?("#{computed_homedir}") }
+          end
+        end # end notifying block
```
I think I'd prefer skipping the `# end ...` comments, since nested `do...end` blocks are a reality in Ruby anyway, everybody should be able to read the code even without the helping comments? If the nesting would make the code _really_ incomprehensible, it's probably a sign that some code should be extracted to a method, lambda, etc?

karmi commented on test/fixtures/cookbooks/elasticsearch_test/recipes/user.rb:
```diff
@@ -0,0 +1,15 @@
+# create user with crazy override options
```
Can we define "crazy" or be more specific here?

karmi commented:
When I run integration tests, I'm getting the following exception:

```
$ time bundle exec rake kitchen:default-centos-65
...
Chef Client failed. 4 resources updated in 24.21197519 seconds
       [2015-01-04T18:17:17+00:00] ERROR: elasticsearch_user[elasticsearch] (elasticsearch::default line 10) had an error: NoMethodError: undefined method `new_resource' for Chef::Resource::ElasticsearchUser
       [2015-01-04T18:17:17+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

karmi commented:
It seems like some Rspec-based gems are not depending properly on the `rspec` gem, because this happens:

```
$ bundle exec rake -T
rake aborted!
LoadError: cannot load such file -- rspec/core/rake_task
/Users/karmi/Playground/ElasticSearch/Varia/cookbook-elasticsearch-new/Rakefile:49:in `<top (required)>'
```

As soon as I add `rspec` to the `group :unit` in the Gemfile, it works.

karmi commented:
When I try to run the RSpec unit tests, I'm getting the following error:

```
$ bundle exec rake spec
/Users/karmi/.rbenv/versions/2.2.0/bin/ruby -I/Users/karmi/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/lib:/Users/karmi/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-support-3.1.2/lib /Users/karmi/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.1.7/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb test/unit
/Users/karmi/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/chefspec-0.0.1/lib/chefspec.rb:1:in `require': cannot load such file -- chef (LoadError)
	from /Users/karmi/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/chefspec-0.0.1/lib/chefspec.rb:1:in `<top (required)>'
	from /Users/karmi/Playground/ElasticSearch/Varia/cookbook-elasticsearch-new/test/unit/spec/spec_helper.rb:3:in `require'
```

martinb3 commented:
I fixed the unit tests and rspec dependency (I guess chefspec doesn't depend on it correctly); I'd recommend clearing out any Gemfile.lock and running those again. Regarding the kitchen tests, I am running them with `bundle exec kitchen test default-centos-65`. I found one mistake there, with a wrong variable name, and fixed it (the error is misleading; the way the DSL is evaluated means that `new_resource.bad_variable` will complain about `new_resource` not existing on the right object when `bad_variable` doesn't exist).

In short, try it again? :) Thanks!

martinb3 commented:
Pushed new version of this branch with Gemfile fixes.
